### PR TITLE
jwt-middleware +tests

### PR DESCRIPTION
**Pull request by Darjusch**
Middleware that checks if the request carries a jwt and if the token is valid.
To test clone the current branch, add the secrets from the chat in the config/test.json, run npx tsc and then npm run docker_test.

### CODE REVIEW

**Review by thatsddr — APPROVED** [review]
The tests pass but there are some vulnerabilities that can't be fixed by just running "npm audit fix" and that probably is something worth taking a look at:
Screen Shot 2021-03-25 at 15 20 20

This is not however something that needs to be done for this pull request.

Another thing I noticed is that there are some "var"s in the code that could be changed with either "const" or "let".

Besides this, the code looks good.

File: docker-compose.yml [hunk 1]
```diff
@@ -4,7 +4,7 @@ services:
     environment:
       - NODE_ENV=dev
     container_name: server
-    image: goodbuy/login
+    image: goodbuy/nodejs-webserver:jwt-verify-request
```
Comment (5h3rr1ll) [inline docker-compose.yml:7]: Naming for the image. Please rename it to `goodbuy/nodejs-webserver` until we have a tag var in place.

File: src/controllers/auth.ts [hunk 2]
```diff
@@ -62,4 +62,18 @@ export const loginUser = (req: Request, res: Response) => {
             return res.status(501).json({ message: "internal server error" })
         }
         );
-}
\ No newline at end of file
+}
+
+export const authenticateToken = (req: any, res: Response, next: NextFunction) => {
```
Comment (5h3rr1ll) [inline src/controllers/auth.ts:67]: Please define a propper type for the request var, `any` is only okay while developing locally.

File: test/auth.js [hunk 3]
```diff
@@ -97,4 +97,46 @@ describe('Authentication', () => {
...
+            let product = {
+                name: "test_product",
+                brand: "test_product",
+                corporation: "test_corp",
+                barcode: "123456789",
+                state: "unverified"
+            }
+            chai.request(server)
+            .post('/product')
+            .set({ "Authorization": `Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImFkbWluQHRlc3QuZGUiLCJpYXQiOjE2MTY1ODI3MjksImV4cCI6MTYxNzE4NzUyOX0.SXWuVkQu1fz0eCfLzN1E93TtZEYsoR60KB_UywiqIQ0`})
```
Comment (5h3rr1ll) [inline test/auth.js:130]: Sure that you want to commit that Bearer-Token?

**Review by 5h3rr1ll — CHANGES_REQUESTED** [review]
Hey, saw a few things, pls check my comments. ✌️

File: test/auth.js (hunk 3 above)
Comment (Darjusch) [reply test/auth.js:130]: this is a manipulated jwt as written in the test it's not valid

**Comment by jwdotpark** [comment]
As anthony said, stick to ES6 seems good idea. Everything works great on my side! 
(FYI not an issue, there was a generic port issue when ```docker_run``` on wsl2 environment could be a problem for windows user to run but it can be easily bypassed)

**Review by jwdotpark — APPROVED** [review]

**Pull request merged**